tests/unit/scrapy/utils/test_apply_apify_settings.py

```diff
@@ -0,0 +1,109 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+import pytest
+from scrapy.settings import Settings
+
+from apify.scrapy.utils import apply_apify_settings
+
+
+@dataclass(frozen=True)
+class TestCase:
+    settings: Settings
+    proxy_config: dict | None
+    expected_output: Settings
+
+
+test_cases = [
```

Comment:
Too DRY for my liking. As someone reading (not writing) the tests I'd probably prefer to have three separate `test_` functions with descriptive names, testing only the values relevant for that particular test case. In my opinion here the use of `parametrize` obfuscates the aim of each test. The fact that we needed a dataclass and a whole standalone list for the test cases IMHO signals that we're pushing something too hard here.

Reply:
Do you think really this approach is so bad?

IMO it seems to me both approaches have pros and cons.

What you're suggesting will lead to a much longer test file and repeating code (also quite long function names, hah).

My approach doesn't have these negatives, however, could be less readable at first glance. I agree with that.

Reply:
I don't think it's _so_ bad and it wouldn't be a reason to reject the PR (if I was in charge). I think DRY doesn't suit situations where you write the code once and read it 50 times, which tests usually are. I just don't see a big benefit of not repeating code here, and it comes with the readability penalty, which IMHO, in tests, has a larger impact. Function names could be:

- `test_apply_apify_settings_no_proxy_config`
- `test_apply_apify_settings_proxy_config`
- `test_apply_apify_settings_custom_settings_and_proxy_config`

Doesn't look so terrible to me. But of course, YMMV.

Reply:
I tried to refactor it as an exercise and ended up with something like this:

```python
def test_apply_apify_settings_takes_default_settings_from_scrapy():
    settings = apply_apify_settings()

    assert settings['DOWNLOADER_MIDDLEWARES'].get('apify.scrapy.middlewares.ApifyRetryMiddleware') == 1000


def test_apply_apify_settings_overrides_scheduler():
    settings = apply_apify_settings(settings=Settings())

    assert settings['SCHEDULER'] == 'apify.scrapy.scheduler.ApifyScheduler'


def test_apply_apify_settings_adds_dataset_pipeline():
    settings = apply_apify_settings(settings=Settings())

    assert settings['ITEM_PIPELINES'] == {'apify.scrapy.pipelines.ActorDatasetPushPipeline': 1000}


def test_apply_apify_settings_adds_proxy_middleware():
    settings = apply_apify_settings(settings=Settings())

    assert settings['DOWNLOADER_MIDDLEWARES'].get('apify.scrapy.middlewares.ApifyHttpProxyMiddleware') == 950


def test_apply_apify_settings_adds_retry_middleware():
    settings = apply_apify_settings(settings=Settings())

    assert settings['DOWNLOADER_MIDDLEWARES'].get('apify.scrapy.middlewares.ApifyRetryMiddleware') == 1000


@pytest.mark.parametrize(
    'proxy_config, expected',
    [
        (None, None),
        ({'useApifyProxy': True, 'apifyProxyGroups': []}, {'useApifyProxy': True, 'apifyProxyGroups': []}),
    ]
)
def test_apply_apify_settings_sets_proxy(proxy_config, expected):
    settings = apply_apify_settings(settings=Settings(), proxy_config=proxy_config)

    assert settings['APIFY_PROXY_SETTINGS'] == expected
```

It's written from the top of my head, I didn't execute the code, so it can contain mistakes. I just wanted to show the approach I had in mind. I iterated towards having a separate test case for each of the changes the function does. The tests are easy to read and simply "explain" each unit tested.

I can imagine one larger "integration" test which takes whole settings and checks what has been applied, if that's necessary.

I omitted the "custom" settings completely, because I realized what we're doing here is putting custom settings to Scrapy's `Settings()`, but that only tests whether Scrapy can apply those. Each of our tests actually uses the part where we pass custom settings in an argument, and none of them is testing settings set to `None`, in which case we load default settings from Scrapy. That part is hard to test. I can imagine approaches to do that, but I don't think they're worth it, I'd leave it as it is.

Reply:
Yeah... agreed and thank you :slightly_smiling_face:. I created a separate test function for every test case. Seems more readable and definitely will be easier to find out where the problem is when one of the test cases fails.

Reply:
I understand the views of both of you and agree that in this case, it's probably better to just write everything out. However, naming test functions for example, is a pain point for many people (including me) - see for example https://github.com/darrenburns/ward, which tries to solve that. But I'm by no means suggesting we adopt that, just wanted to point out that you can typically think of more useful test cases than fitting names.

Reply:
Yup, writing (py)test function names is a bit annoying and while I wasn't really amused by all of the [mocha mess](https://mastodonczech.cz/@iamkale@infosec.exchange/111801113019589560) back in the early node.js days when coming from Python, I think the decision to describe test cases in plain text was a good decision.

Copilot to the rescue 😄 Today we can just write the test first and let AI to suggest the function name.